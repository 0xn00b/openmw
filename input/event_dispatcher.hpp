#ifndef _INPUT_EVENT_DISPATCHER_H
#define _INPUT_EVENT_DISPATCHER_H

#include "dispatch_map.hpp"
#include <boost/function.hpp>

namespace Input {

  /**
     The EventDispatcher translates an input event (as given by an
     identifying index and an optional void pointer) into an output
     event function call.
  */
  class EventDispatcher
  {
    DispatchMap map;

    /*
      The event callback function that is called for all events. The
      first parameter is the input event. The second parameter is the
      resolved output event. The third is an optional user-defined
      parameter passed to call().
    */
    typedef boost::function<void(int,int,void*)> EventCallback;

    EventCallback callback;

    // Carpal-tunnel prevention
    typedef DispatchMap::OutList _OL;

  public:

    /// Create an event binding connection
    void bind(int in, int out)
    { map.bind(in,out); }

    /// Dissolve an event binding connection
    void unbind(int in, int out)
    { map.unbind(in, out); }

    /// Check if a given input is bound to anything
    bool isBound(int in)
    { return map.isBound(in); }

    /// Register the callback you want to use to handle events.
    void setCallback(EventCallback cb)
    { callback = cb; }

    /**
       Instigate an event.

       This will look up the input event number (first parameter), and
       call the event callback for each output number associated with
       (bound to) that input.

       The optional second paramter is also passed to the callback.

       If no output is bound to the given event number, the callback
       is never called.
    */
    void call(int event, void *p = NULL)
    {
      // You have to set the callback before using call().
      assert(!callback.empty());

      // Not bound? Exit.
      if(!isBound(event)) return;

      // Dispatch to all events.
      const _OL &list = map.getList(event);
      for(_OL::const_iterator it = list.begin();
          it != list.end(); it++)
        callback(event, *it, p);
    }
  };

}
#endif
