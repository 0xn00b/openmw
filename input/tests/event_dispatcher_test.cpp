#include <iostream>
using namespace std;

#include "../event_dispatcher.hpp"

using namespace Input;

void callback(int in, int out, void *p)
{
  cout << "    Got event: in=" << in << " out=" << out << endl;
}

EventDispatcher dsp;

void callAll()
{
  cout << "\nDuty calls:\n";
  for(int i=1; i<5; i++)
    {
      cout << "  Calling event " << i << ":\n";
      dsp.call(i);
    }
}

int main()
{
  cout << "Testing the event dispatcher\n";

  dsp.setCallback(&callback);

  callAll();

  dsp.bind(2,1);
  dsp.bind(1,10);
  dsp.bind(14,-12);
  dsp.bind(2,-137);

  callAll();

  dsp.unbind(1,8);
  dsp.unbind(1,10);
  dsp.unbind(2,-137);
  dsp.unbind(2,1);

  callAll();

  dsp.bind(3, 19);
  dsp.bind(4, 18);
  dsp.bind(4, 18);

  callAll();

  return 0;
}
