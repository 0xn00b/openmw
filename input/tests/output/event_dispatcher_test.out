Testing the event dispatcher

Duty calls:
  Calling event 1:
  Calling event 2:
  Calling event 3:
  Calling event 4:

Duty calls:
  Calling event 1:
    Got event: in=1 out=10
  Calling event 2:
    Got event: in=2 out=-137
    Got event: in=2 out=1
  Calling event 3:
  Calling event 4:

Duty calls:
  Calling event 1:
  Calling event 2:
  Calling event 3:
  Calling event 4:

Duty calls:
  Calling event 1:
  Calling event 2:
  Calling event 3:
    Got event: in=3 out=19
  Calling event 4:
    Got event: in=4 out=18
